specs/discussions/2025-05-segments-global-history.md

```diff
@@ -0,0 +1,320 @@
+
+> Note: This is a high-level overview of how re-orgs are handled and why they shouldn't impact
+> greatly the operation of shards. Let me know if additional details are needed here and I can
+> elaborate a bit more on it.
+
+- When, e.g. an intermediate shard proposes a new block `blkA`, this block will have a reference to
+  a valid block from the history of the beacon chain.
+- `blkA` will be submitted to the intermediate's shard parent, in this case the beacon chain, inside
+  an `IntermediateShardBlockInformation` data structure.
+- This process will be repeated for every new block in the shard: `blkB`, `blkC`, etc.
+- Verifying the validity of a shard block is done by making the regular block validity verification
+  and all the additional verifications imposed by the hierarchical consensus like checking that the
+  referenced beacon chain block is valid. You can already glimpse here how we are not only embedding
+  in the longest chain rule local consensus verifications, but also logic involving the overall
+  hierarchical consensus operation.
+- So far so good. However, all the shards in the system are running a probabilistic consensus. What
+  happens if there is a chain re-organisation and a heaviest chain surfaces replacing the current
+  longest-chain? Depending on the shard suffering the re-org, the impact is different. Fortunately,
+  all of them can be handled quite gracefully through the longest-chain rule.
+
+  - Upward reorgs, meaning leaf or intermediate shards that suffer a re-org, are easily detected. As
+    blocks are being submitted to the parent, the parent is able to identify that there is a
+    heaviest chain that needs to replace its current view of the child shard. This requires no
+    immediate action from the parent apart from this update of the parent's view of the heaviest
+    chain.
+  - Reorgs from intermediate shards do not involve any immediate action from its child shards. The
+    newest heaviest chain may change the way (and specific blocks) where the information about the
+    child shard is being included into the parent chain and propagated to the parent chain, but this
+    shouldn't have any additional impact.
+  - Finally, beacon chain re-orgs are the only onces that may trigger actions in the lower levels of
+    the hierarchy. When there is a heaviest chain in the beacon chain replacing the current longest
+    chain, this can trigger a re-org in some shards of the lower levels of the hierarchy that were
+    following the previous longest chain, as their beacon chain references for some of their blocks
+    may be deemed invalid. This event is also handled gracefully by the longest-chain rule. The
+    farmer entitled to create the next block in the child shard after the beacon re-org will build
+    the new block on top of the latest block in the child shard with a valid beacon chain block
+    reference after the re-org, continuing in this way constructing a valid chain in the shard.
```

Comment:
This essentially boils down to a much shorter explanation that looks something like this:
```suggestion
  - In case of beacon chain reorg, all blocks of the lower level shards that reference stale beacon chain blocks automatically become stale too. Lower-level shard blocks must always point to beacon chain blocks from the canonical branch.
```